### PR TITLE
Created LoadTest project to reproduce StackOverflow exception

### PR DESCRIPTION
**Pull request by Havunen**
I created new project LoadTest to reproduce this Github issue https://github.com/dadhi/DryIoc/issues/139
Its console application which resolves multiple types and fails to different exceptions based on Container configs in program.cs

I hope this can be used to fix the stackoverflow issue and hopefully demostrate locking and high memory consumption under heavy load.

### CODE REVIEW

**Comment by Havunen** [comment]
You can play with `threadCount` and `iterations` variables to tune the load for your PC.

File: playground/LoadTest/LoadTest.csproj [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net461</TargetFramework>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <NoWarn>1701;1702;0003;CS0108;CS0109;CS0169;CS0067;IDE0060;IDE0022;IDE0052;IDE0044;IDE0051;IDE0059</NoWarn>
```
Comment (Havunen) [inline playground/LoadTest/LoadTest.csproj]: I ignored lot of warnings because they are not relevant for this project (unused properties / fields, etc )

**Comment by Havunen** [comment]
The files inside "gen" folder are just types used for registration / resolve process and contain no real code

**Comment by dadhi** [comment]
Thanks, very much appreciated.

Will check.

**Comment by Havunen** [comment]
I analysed load test with dotMemory and it seems a almost all of the memory traffic comes from System.Linq.Expressions and System.Reflection which System.Linq.Expression uses behind the scenes. The memory flow starts from `CompileToFactoryDelegate` 

https://github.com/dadhi/DryIoc/blob/master/src/DryIoc/Container.cs#L2866-L2870

Is there some reason why all Controllers need to be compiled to lamda expression over and over again?

**Comment by Havunen** [comment]
I found that adding cache around expression compilation helps quite a bit with memory traffic, to give an idea it could be something like this:

```c#
        private static ConcurrentDictionary<int, FactoryDelegate> TypeToDelegateCache = new ConcurrentDictionary<int, FactoryDelegate>();

        /// <summary>First wraps the input service expression into lambda expression and
        /// then compiles lambda expression to actual <see cref="FactoryDelegate"/> used for service resolution.</summary>
        public static FactoryDelegate CompileToFactoryDelegate(this Expression expression,
            bool useFastExpressionCompiler = false)
        {
            FactoryDelegate del;
            int hash = expression.Type.GetHashCode();
            if (TypeToDelegateCache.TryGetValue(hash, out del))
            {
                return del;
            }

            expression = expression.NormalizeExpression();

            // Optimization for constants
            if (expression is ConstantExpression ce)
                return ce.Value.ToFactoryDelegate;

            if (useFastExpressionCompiler)
            {
                var factoryDelegate = FastExpressionCompiler.LightExpression.ExpressionCompiler.TryCompile<FactoryDelegate>(
                    expression, _factoryDelegateParamExprs, _factoryDelegateParamTypes, typeof(object));
                if (factoryDelegate != null)
                    return factoryDelegate;
            }

            // fallback for platforms when FastExpressionCompiler is not supported,
            // or just in case when some expression is not supported (did not found one yet)
            var lambdaExpr = Lambda<FactoryDelegate>(expression, _factoryDelegateParamExprs);

#if SUPPORTS_FAST_EXPRESSION_COMPILER
            del = lambdaExpr.ToLambdaExpression().Compile();
#else
            del = lambdaExpr.Compile();
#endif
            TypeToDelegateCache.TryAdd(hash, del);
            return del;
        }
```

**Comment by Havunen** [comment]
After adding cache there is still many System.Linq.Expression ( NewExpression ) objects

**Comment by Havunen** [comment]
Btw about the original issue: I noticed it can be reproduced without multi threading also. This should make it easier to fix. 

Its reproducing this way too:
```c#
var iterations = 1000;
var i = 0;

var controllers = controllerTypes;
var controllersCount = controllers.Length;

for (var j = 0; j < iterations; j++)
{
	for (var k = 0; k < controllersCount; k++)
	{
		// Simulate WebAPI loop, open scope resolve and repeat
		using (var scope = container.OpenScope(Reuse.WebRequestScopeName))
		{
			scope.Resolve(controllers[k]);
		}
	}
}


var stopWatch = new Stopwatch();
stopWatch.Start();

stopWatch.Stop();
// Get the elapsed time as a TimeSpan value.
var ts = stopWatch.Elapsed;
Console.WriteLine("-- Load Test Result --");
Console.WriteLine($"{ts.Hours:00}:{ts.Minutes:00}:{ts.Seconds:00}.{ts.Milliseconds / 10:00}");
```

**Comment by dadhi** [comment]
This is good catch.

**Comment by Havunen** [comment]
@dadhi did you have chance to look at this issue yet? Were you able to reproduce the exception?

**Comment by dadhi** [comment]
Not yet, I was thinking about the usual suspects to check first.

At the moment I am integrating a new version of FEC in separate branch which might address (hopefully) some of suspects.

Will do a detailed check after your PR is merged.

**Comment by dadhi** [comment]
> Is there some reason why all Controllers need to be compiled to lamda expression over and over again?

It should be compiled on the first resolve only.

**Pull request merged**